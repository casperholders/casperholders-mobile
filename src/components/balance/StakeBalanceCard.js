import formatCasperAmount from '@/helpers/formatCasperAmount';
import truncateInMiddle from '@/helpers/truncateInMiddle';
import useAdapter from '@/hooks/auth/useAdapter';
import ReadOnlyAdapter from '@/services/signers/readOnlyAdapter';
import { StyleSheet } from 'react-native';
import { Button, Caption, Card, Paragraph, useTheme } from 'react-native-paper';

export default function (
  { navigation, validator, staked, formattedPercentOfTotal, delegationRate },
) {
  const adapter = useAdapter();
  const readOnly = adapter.constructor.ID === ReadOnlyAdapter.ID;
  const theme = useTheme();
  const description = `${formattedPercentOfTotal}% of your staked funds Íûè Validator ${truncateInMiddle(validator)}`;

  return (
    <Card style={styles.card}>
      <Card.Content>
        <Paragraph>
          {formatCasperAmount(staked)}
        </Paragraph>
        <Caption>
          {description}
        </Caption>
      </Card.Content>
      <Card.Actions style={styles.actions}>
        <Button
          style={styles.actionsStake}
          color={theme.colors.text}
          onPress={() => navigation.jumpTo('OperationsTab', {
            screen: 'Stake',
            address: validator,
            initial: false,
          })}
          disabled={readOnly}
        >
          Stake
        </Button>
        <Button
          color={theme.colors.text}
          onPress={() => navigation.jumpTo('OperationsTab', {
            screen: 'Unstake',
            address: validator,
            initial: false,
          })}
          disabled={readOnly}
        >
          Unstake
        </Button>
      </Card.Actions>
    </Card>
  );
}

const styles = StyleSheet.create({
  card: {
    marginLeft: 12,
    marginBottom: 12,
  },
  actions: {
    alignItems: 'flex-end',
    justifyContent: 'flex-end',
  },
  actionsStake: {
    marginRight: 8,
  },
});
