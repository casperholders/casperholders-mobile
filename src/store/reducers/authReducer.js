import signerManager from '@/services/newSigners/signerManager';
import { createSlice } from '@reduxjs/toolkit';

const authSlice = createSlice({
  name: 'auth',
  initialState: {
    connected: false,
    adapterId: undefined,
    options: undefined,
  },
  reducers: {
    connect: (state, action) => {
      state.adapterId = action.payload.adapterId;
      state.options = action.payload.options;
      state.connected = true;
    },
    disconnect: (state) => {
      state.connected = false;
      state.options = undefined;
      state.adapterId = undefined;
    },
  },
});

export const { connect, disconnect } = authSlice.actions;

export const selectConnected = (state) => state.auth.connected;

export const selectAdapter = (state) => signerManager.adapter(state.auth.adapterId);

export const selectOptions = (state) => state.auth.options;

export const selectPublicKey = (state) => selectAdapter(state).computePublicKey(selectOptions(state));

export default authSlice.reducer;
