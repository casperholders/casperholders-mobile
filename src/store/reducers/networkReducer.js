import MainnetAdapter from '@/services/networks/mainnetAdapter';
import NetworkManager from '@/services/networks/networkManager';
import { createSlice } from '@reduxjs/toolkit';

/**
 * NetworkReducer manage the network state inside the app
 * @type {Slice<{networkId: string}, {setNetwork: reducers.setNetwork}, string>}
 */
const networkSlice = createSlice({
  name: 'network',
  initialState: {
    networkId: MainnetAdapter.ID,
  },
  reducers: {
    setNetwork: (state, action) => {
      state.networkId = action.payload.network;
    },
  },
});

export const { setNetwork } = networkSlice.actions;

export const selectNetwork = (state) => {
  return NetworkManager.adapter(state.network.networkId);
};

export default networkSlice.reducer;
