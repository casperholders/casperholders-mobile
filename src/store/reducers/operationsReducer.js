import { createSlice } from '@reduxjs/toolkit';

const operationsSlice = createSlice({
  name: 'operations',
  initialState: {
    pendingNotificationId: undefined,
    deployResultsByHash: {},
  },
  reducers: {
    setDeployResult: (state, action) => {
      state.deployResultsByHash[action.payload.deployResult.hash] = action.payload.deployResult;
    },
    unsetDeployResult: (state, action) => {
      const lowerHash = action.payload.deployResult.hash.toLowerCase();
      if (lowerHash in state.deployResultsByHash) {
        delete state.deployResultsByHash[lowerHash];
      }
    },
    setPendingNotificationId: (state, action) => {
      state.pendingNotificationId = action.payload.notificationId;
    },
  },
});

export const {
  setDeployResult,
  unsetDeployResult,
  setPendingNotificationId,
} = operationsSlice.actions;

export default operationsSlice.reducer;
