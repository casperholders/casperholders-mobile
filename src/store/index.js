import authReducer from '@/store/reducers/authReducer';
import networkReducer from '@/store/reducers/networkReducer';
import operationsReducer from '@/store/reducers/operationsReducer';
import { configureStore } from '@reduxjs/toolkit';

export default configureStore({
  reducer: {
    auth: authReducer,
    operations: operationsReducer,
    network: networkReducer,
  },
});
