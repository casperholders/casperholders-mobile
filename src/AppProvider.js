import { getStore } from '@/store';
import theme from '@/theme';
import Big from 'big.js';
import { Provider as PaperProvider } from 'react-native-paper';
import { Provider as ReduxProvider } from 'react-redux';
import toFormat from 'toformat';

toFormat(Big);

export default function AppProvider({ children }) {
  return (
    <ReduxProvider store={getStore()}>
      <PaperProvider theme={theme}>
        {children}
      </PaperProvider>
    </ReduxProvider>
  );
}
