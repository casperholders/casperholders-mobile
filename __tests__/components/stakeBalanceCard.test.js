import AppProvider from '@/AppProvider';
import StakeBalanceCard from '@/components/balance/StakeBalanceCard';
import LocalAdapter from '@/services/signers/localAdapter';
import { getStore, resetStore } from '@/store';
import { connect } from '@/store/reducers/authReducer';
import { TEST_LOCAL_SIGNER_KEY } from '@env';
import { render } from '@testing-library/react-native';

beforeEach(() => {
  resetStore();
});

jest.setTimeout(500000);

describe('Stake balance card', () => {
  test('should render stake balance card', () => {
    getStore().dispatch(
      connect({
        adapterId: LocalAdapter.ID,
        options: { privateKey: TEST_LOCAL_SIGNER_KEY },
      }),
    );

    const validator = {
      png: '',
      name: 'CasperHolders',
      publicKey: '0124bfdae2ed128fa5e4057bc398e4933329570e47240e57fc92f5611a6178eba5',
      group: 'Active',
      delegation_rate: '0',
      staked_amount: '253.00',
    };

    const { debug, queryByText, toJSON } = render(<StakeBalanceCard
      validator={'CasperHolders'}
      image={'https://node.casperholders.com/.well-known/casper/squarelogo256.png'}
      staked={100}
      formattedPercentOfTotal={'1'}
      delegationRate={'0'}
    />, {
      wrapper: AppProvider,
    });
    expect(queryByText(/Validator CasperHolders/i)).toBeTruthy();
    expect(queryByText(/1% of your staked funds - 0% fee/i)).toBeTruthy();

    expect(toJSON()).toMatchSnapshot();
  });
});