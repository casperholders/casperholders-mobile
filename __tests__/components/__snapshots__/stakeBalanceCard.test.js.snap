// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stake balance card should render stake balance card 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "rgba(13, 14, 53, 1)",
        "borderColor": "rgba(255, 255, 255, 0.12)",
        "borderRadius": 8,
        "elevation": 1,
        "marginBottom": 12,
        "marginLeft": 12,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0.5,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 0.75,
      }
    }
  >
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flexGrow": 1,
          "flexShrink": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "paddingHorizontal": 16,
            },
            Object {
              "paddingVertical": 16,
            },
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignSelf": "center",
              "marginBottom": "auto",
              "paddingLeft": 0,
              "paddingRight": 12,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 17.5,
                  "height": 35,
                  "width": 35,
                },
                Object {
                  "backgroundColor": "white",
                },
              ]
            }
          >
            <Image
              source={
                Object {
                  "uri": "https://node.casperholders.com/.well-known/casper/squarelogo256.png",
                }
              }
              style={
                Object {
                  "borderRadius": 17.5,
                  "height": 35,
                  "width": 35,
                }
              }
            />
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(255, 255, 255, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 14,
                        "letterSpacing": 0.25,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
            >
              Validator 
              CasperHolders
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(255, 255, 255, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 14,
                        "letterSpacing": 0.25,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
            >
              - CSPR
            </Text>
            <Text
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(255, 255, 255, 0.54)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 12,
                        "letterSpacing": 0.4,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      undefined,
                    ],
                  ],
                ]
              }
            >
              1% of your staked funds - 0% fee
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-end",
                "flexDirection": "row",
                "justifyContent": "flex-end",
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "transparent",
                  "borderColor": "transparent",
                  "borderRadius": 8,
                  "borderStyle": "solid",
                  "borderWidth": 0,
                  "elevation": 0,
                  "marginRight": 8,
                  "minWidth": 64,
                  "shadowColor": "#000000",
                  "shadowOffset": Object {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                }
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  Object {
                    "disabled": false,
                  }
                }
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 8,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    numberOfLines={1}
                    selectable={false}
                    style={
                      Array [
                        Object {
                          "color": "#ffffff",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "letterSpacing": 1,
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          undefined,
                          Object {
                            "textTransform": "uppercase",
                          },
                          Object {
                            "color": "#ffffff",
                            "fontFamily": "System",
                            "fontWeight": "500",
                          },
                          Object {
                            "fontFamily": "System",
                            "fontWeight": "500",
                          },
                          undefined,
                        ],
                      ]
                    }
                  >
                    Stake
                  </Text>
                </View>
              </View>
            </View>
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "transparent",
                  "borderColor": "transparent",
                  "borderRadius": 8,
                  "borderStyle": "solid",
                  "borderWidth": 0,
                  "elevation": 0,
                  "minWidth": 64,
                  "shadowColor": "#000000",
                  "shadowOffset": Object {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                }
              }
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  Object {
                    "disabled": false,
                  }
                }
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 8,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    numberOfLines={1}
                    selectable={false}
                    style={
                      Array [
                        Object {
                          "color": "#ffffff",
                          "fontFamily": "System",
                          "fontWeight": "400",
                        },
                        Object {
                          "textAlign": "left",
                        },
                        Array [
                          Object {
                            "letterSpacing": 1,
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          undefined,
                          Object {
                            "textTransform": "uppercase",
                          },
                          Object {
                            "color": "#ffffff",
                            "fontFamily": "System",
                            "fontWeight": "500",
                          },
                          Object {
                            "fontFamily": "System",
                            "fontWeight": "500",
                          },
                          undefined,
                        ],
                      ]
                    }
                  >
                    Unstake
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
